docs/writing-controllers.md

```diff
@@ -0,0 +1,406 @@
+
+✅
+
+```typescript
+class FooController extends BaseController {
+  constructor({
+    messenger,
+  }, {
+    messenger: FooControllerMessenger
+  }) {
+    messenger.subscribe(
+      'BarController:stateChange',
+      ([baz, qux, foz]) => {
+        doSomethingWith(baz, qux, foz);
+      },
+      (state) => [state.baz, state.qux, state.foz];
+    );
+  }
+}
+```
+
+## Don't represent data in state multiple ways
+
+Be mindful that state is part of the controller's API just as much as its methods are, and therefore it should also be as concise as possible. A piece of data should only be represented one way; if there are multiple instances of the same objects in separate parts of state, those instances should be extracted to a new state property, and the higher level forms should be removed and derived by the client.
+
+### Also see
+
+- ["Keep state minimal and derive additional values"](https://redux.js.org/style-guide/#keep-state-minimal-and-derive-additional-values) in the Redux Style Guide.
+
+## Remove getters in favor of direct state access
+
+The `state` of a controller should be simple enough such that consumers are able to access it directly rather than using getters:
+
+🚫
+
+```typescript
+class FooController extends BaseController {
+  getActiveAccounts() {
+    return this.state.accounts.filter((account) => account.isActive);
+  }
+
+  get inactiveAccounts() {
+    return this.state.accounts.filter((account) => !account.isActive);
+  }
+}
+
+// Then, later...
+const fooController = new FooController();
+
+for (const account of fooController.getActiveAccounts()) {
+  // ...
+}
+
+for (const account of fooController.inactiveAccounts) {
+  // ...
+}
+```
+
+✅
+
+```typescript
+class FooController extends BaseController {}
+
+// Then, later...
+const fooController = new FooController();
+
+const activeAccounts = fooController.state.accounts.filter(
+  (account) => account.isActive,
+);
+for (const account of activeAccounts) {
+  // ...
+}
+
+const inactiveAccounts = fooController.state.accounts.filter(
+  (account) => !account.isActive,
+);
+for (const account of inactiveAccounts) {
+  // ...
+}
+```
+
+Instead of getters, try selectors:
+
+💡
+
+```typescript
+// === foo-controller ===
+
+type FooControllerState = {
+  // ...
+};
+
+export fooControllerSelectors = {
+  activeAccounts: (state: FooControllerState) => {
+    return state.accounts.filter(
+      (account) => account.isActive,
+    )
+  },
+  inactiveAccounts: (state: FooControllerState) => {
+    return state.accounts.filter(
+      (account) => !account.isActive,
+    )
+  },
+};
+
+export class FooController extends BaseController {
+}
+
+// === Client ==
+
+import { select } from "@metamask/base-controller";
+import { FooController, fooControllerSelectors } from "@metamask/foo-controller";
+
+const fooController = new FooController();
+
+const [activeAccounts, inactiveAccounts] = select(
```

Comment:
It should probably be `fooController.state` one line down from here, rather than `fooController`. We'll need to select from the state. But aside from that, this looks basically like I had imagined as well!

Reply:
I ended up updating this example to use `reselect`.